IPA.Loader/Config/ConfigProviders/IniConfigProvider.cs

```diff
@@ -0,0 +1,268 @@
+
+            var fileInfo = new FileInfo(Filename);
+            if (fileInfo.Exists)
+            {
+                try
+                {
+                    var parser = new FileIniDataParser();
+                    parser.Parser.Configuration.CaseInsensitive = true;
+
+                    _iniData = parser.ReadFile(fileInfo.FullName);
+                }
+                catch (Exception e)
+                {
+                    Logger.config.Error($"Error parsing INI in file {Filename}; resetting to empty INI");
+                    Logger.config.Error(e);
+
+                    _iniData = new IniData();
+                }
+            }
+            else
+            {
+                Logger.config.Debug($"File {fileInfo.FullName} doesn't exist");
+                _iniData = new IniData();
+            }
+
+            InMemoryChanged = true;
+        }
+
+
+        // This is basically trying to deserialize from INI data to a config object
+        public T Parse<T>()
+        {
+            // Create an instance of the config object to return
+            T configObj = Activator.CreateInstance<T>();
+
+            // Get a list of the fields declared in the config object
+            Type configObjType = typeof(T);
+
+            // Create a dictionary to record which fields are found in the class files
+            Dictionary<string, FieldInfo> classConfigField = new Dictionary<string, FieldInfo>();
+
+            // This goes through each field of the class to set values
+            // if found in the configuration file
+            foreach (FieldInfo field in configObjType.GetFields())
+            {
+                Type fieldType = field.FieldType;
+
+                // If thie field is an object, loop through its fields ("subfields")
+                if (Type.GetTypeCode(fieldType) == TypeCode.Object)
+                {
+
+                    // Get the sub object value from the config object
+                    object configObjSubObj = field.GetValue(configObj);
+
+                    foreach (FieldInfo subField in fieldType.GetFields())
+                    {
+
+                        // If the INI file has a section/key pair corresponding to the field/subfield,
+                        // set the subfield value and store the field info in dictionary
+                        if (_iniData.Sections.ContainsSection(field.Name) && _iniData[field.Name].ContainsKey(subField.Name))
+                        {
+                            SetFieldValue(subField, configObjSubObj, _iniData[field.Name][subField.Name]);
+                            string fieldName = field.Name + "." + subField.Name;
+                            classConfigField[fieldName.ToUpper()] = subField;
+                        }
+                        else
+                        {
+                            Logger.config.Debug($"{field.Name}.{subField.Name} doesn't have a configuration value! Keeping existing value {subField.GetValue(configObjSubObj)}");
```

Comment:
This should maybe be a `Warn` or `Notice`?

Reply:
Yeah, I was using Debug for all my Log statements but it probably makes sense to make a lot of these Warn or Notice. Do you have a set of guidelines for deciding between the two?

Reply:
Notices generally mean something that's probably important, but not really an issue. Warn, on the other hand, means there *is* an issue, but its not big enough to actually cause many real-world problems. Its just a subset of [syslog](https://en.wikipedia.org/wiki/Syslog).